### PR TITLE
SMAPI integration

### PR DESCRIPTION
**Pull request by chris-ramon**
#### Overview
- [x] `/v1/validateSource`: Adds support skills syncing between SMAPI and Firebase.
- [x] `/v1/skillAuthorized`: New endpoint for checking skill auth permissions for given user token and invocation name.
- [x] `/linkAVS`: Adds support for calling `virtual-device.bespoken.io` to retrieve for a user its SMAPI access token, and then saved to Firebase.

#### Test plan
- [x] Unit test
- [x] Manual test

#### Action required
- Merge along `dashboard` PR: https://github.com/bespoken/dashboard/pull/466

### CODE REVIEW

File: src/main/services/v1/SkillAuthorized.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                        }
+                        for (const locale in skill.interactionModels) {
+                            const interactionModel = skill.interactionModels[locale];
+                            if (!interactionModel
+                                || !interactionModel.languageModel
+                                || !interactionModel.languageModel.invocationName) {
+                                continue
+                            }
+                            if (interactionModel.languageModel
+                                    .invocationName === invocation_name.trim().toLowerCase()
```
Comment (jkelvie) [inline src/main/services/v1/SkillAuthorized.ts]: invocation_name.includes(interactionModel.languageModel.invocationName)

File: src/main/firebase/FirebaseController.ts [hunk 2]
```diff
@@ -311,6 +324,11 @@ export class FirebaseDBUser implements FirebaseUserObj {
         teamObj[id].enableNotifications = enableNotifications;
         return this.myRef.child("team").set(teamObj);
     }
+
+    setSkill(skillId: string, vendor: any) {
+        const data: any = {roles: vendor.roles};
+        return this.myRef.child("skills/" + normalizeSkillID(skillId)).update(data);
```
Comment (jkelvie) [inline src/main/firebase/FirebaseController.ts]: Here there is no leading slash, but above there is....does this definitely work?

File: src/main/services/v1/PostValidateSource.ts [hunk 3]
```diff
@@ -103,6 +103,7 @@ export function postValidateSource(db: Admin.database.Database):
         }
         setTimeout(async () => {
```
Comment (jkelvie) [inline src/main/services/v1/PostValidateSource.ts]: Why are you using setTimeout? Why not just put this code into it's own routine, and then call it with an empty done block, like this:
https://github.com/skillbotio/server/blob/master/src/UserSession.ts#L70

Comment (chris-ramon) [reply src/main/services/v1/PostValidateSource.ts]: good call!, addressed on: https://github.com/bespoken/source-api/pull/33/commits/8ddbe650698c1051e4eb570ed2f3ce8b18b94500#diff-2ef49e17a3b0fe9b55949152909d6382R111

File: src/main/services/v1/PostValidateSource.ts [hunk 4]
```diff
@@ -79,58 +79,8 @@ export function postValidateSource(db: Admin.database.Database):
...
-                }
-                if (err === InvalidSilentEchoTokenError) {
-                    const validationResults = reauthenticateHTML();
-                    return triggerEvent({validationResults}, "results", eventChannel(ctx));
-                }
-                const validationResults = UnknownError.message;
-                return triggerEvent({validationResults}, "results", eventChannel(ctx));
-            }
-        });
+        setTimeout(() => syncSkills(db, user_id, sm_api_access_token, token,
```
Comment (jkelvie) [inline src/main/services/v1/PostValidateSource.ts]: Looks good overall, but why do the setTimeout?

Comment (jkelvie) [reply src/main/services/v1/PostValidateSource.ts]: Never mind, I see that change was made on the next pull request.

👍

**Pull request merged**